File: services/ml-service/src/handlers/training.rs
use axum::{Json, extract::State};
use chrono::Utc;
use serde_json::{Value, json};
use sqlx::{FromRow, query_as, query_scalar};
use uuid::Uuid;

use crate::{
    AppState,
    domain::{interop, training},
    models::{
        model_version::ModelVersion,
        training_job::{CreateTrainingJobRequest, ListTrainingJobsResponse, TrainingJob},
    },
};

use super::{ServiceResult, bad_request, db_error, deserialize_json, to_json};

#[derive(Debug, FromRow)]
struct TrainingJobRow {
    id: Uuid,
    experiment_id: Option<Uuid>,
    model_id: Option<Uuid>,
    name: String,
    status: String,
    dataset_ids: Value,
    training_config: Value,
    hyperparameter_search: Value,
    objective_metric_name: String,
    trials: Value,
    best_model_version_id: Option<Uuid>,
    submitted_at: chrono::DateTime<Utc>,
    started_at: Option<chrono::DateTime<Utc>>,
    completed_at: Option<chrono::DateTime<Utc>>,
    created_at: chrono::DateTime<Utc>,
}

#[derive(Debug, FromRow)]
struct ModelVersionRow {
    id: Uuid,
    model_id: Uuid,
    version_number: i32,
    version_label: String,
    stage: String,
    source_run_id: Option<Uuid>,
    training_job_id: Option<Uuid>,
    hyperparameters: Value,
    metrics: Value,
    artifact_uri: Option<String>,
    schema: Value,
    created_at: chrono::DateTime<Utc>,
    promoted_at: Option<chrono::DateTime<Utc>>,
}

fn to_training_job(row: TrainingJobRow) -> TrainingJob {
    let external_training = interop::tracking_source_from_training_config(&row.training_config);
    TrainingJob {
        id: row.id,
        experiment_id: row.experiment_id,
        model_id: row.model_id,
        name: row.name,
        status: row.status,
        dataset_ids: deserialize_json(row.dataset_ids),
        training_config: row.training_config,
        hyperparameter_search: row.hyperparameter_search,
        objective_metric_name: row.objective_metric_name,
        trials: deserialize_json(row.trials),
        best_model_version_id: row.best_model_version_id,
        external_training,
        submitted_at: row.submitted_at,
        started_at: row.started_at,
        completed_at: row.completed_at,
        created_at: row.created_at,
    }
}

#[allow(dead_code)]
fn _to_model_version(row: ModelVersionRow) -> ModelVersion {
    ModelVersion {
        id: row.id,
        model_id: row.model_id,
        version_number: row.version_number,
        version_label: row.version_label,
        stage: row.stage,
        source_run_id: row.source_run_id,
        training_job_id: row.training_job_id,
        hyperparameters: row.hyperparameters,
        metrics: deserialize_json(row.metrics),
        artifact_uri: row.artifact_uri,
        model_adapter: interop::model_adapter_from_schema(&row.schema),
        registry_source: interop::registry_source_from_schema(&row.schema),
        external_tracking: interop::tracking_source_from_schema(&row.schema),
        schema: row.schema,
        created_at: row.created_at,
        promoted_at: row.promoted_at,
    }
}

pub async fn list_training_jobs(
    State(state): State<AppState>,
) -> ServiceResult<ListTrainingJobsResponse> {
    let rows = query_as::<_, TrainingJobRow>(
        r#"
        SELECT
            id,
            experiment_id,
            model_id,
            name,
            status,
            dataset_ids,
            training_config,
            hyperparameter_search,
            objective_metric_name,
            trials,
            best_model_version_id,
            submitted_at,
            started_at,
            completed_at,
            created_at
        FROM ml_training_jobs
        ORDER BY submitted_at DESC, created_at DESC
        "#,
    )
    .fetch_all(&state.db)
    .await
    .map_err(|cause| db_error(&cause))?;

    Ok(Json(ListTrainingJobsResponse {
        data: rows.into_iter().map(to_training_job).collect(),
    }))
}

pub async fn create_training_job(
    State(state): State<AppState>,
    Json(body): Json<CreateTrainingJobRequest>,
) -> ServiceResult<TrainingJob> {
    if body.name.trim().is_empty() {
        return Err(bad_request("training job name is required"));
    }

    let objective_metric_name = body
        .objective_metric_name
        .unwrap_or_else(|| "accuracy".to_string());
    let search = body.hyperparameter_search.unwrap_or_else(|| json!({}));
    let base_training_config = if body.training_config.is_null() {
        json!({ "engine": "tabular-logistic" })
    } else {
        body.training_config
    };
    let resolved_training_config = interop::merge_training_config_with_external(
        base_training_config,
        body.external_training.as_ref(),
    );
    let execution = training::execute_training(
        &resolved_training_config,
        Some(&search),
        &objective_metric_name,
    )
    .map_err(bad_request)?;
    let dataset_ids_for_schema = body.dataset_ids.clone();
    let training_config_for_schema = resolved_training_config.clone();
    let search_for_schema = search.clone();
    let objective_metric_name_for_schema = objective_metric_name.clone();
    let now = Utc::now();
    let job_id = Uuid::now_v7();

    let mut best_model_version_id = None;

    if body.auto_register_model_version {
        if let (Some(model_id), Some(best_hyperparameters)) =
            (body.model_id, execution.best_hyperparameters.clone())
        {
            let next_version_number = query_scalar::<_, i32>(
                "SELECT COALESCE(MAX(version_number), 0) + 1 FROM ml_model_versions WHERE model_id = $1",
            )
            .bind(model_id)
            .fetch_one(&state.db)
            .await
            .map_err(|cause| db_error(&cause))?;

            let version_row = query_as::<_, ModelVersionRow>(
                r#"
                INSERT INTO ml_model_versions (
                    id,
                    model_id,
                    version_number,
                    version_label,
                    stage,
                    source_run_id,
                    training_job_id,
                    hyperparameters,
                    metrics,
                    artifact_uri,
                    schema,
                    promoted_at
                )
                VALUES ($1, $2, $3, $4, 'candidate', NULL, $5, $6, $7, $8, $9, NULL)
                RETURNING
                    id,
                    model_id,
                    version_number,
                    version_label,
                    stage,
                    source_run_id,
                    training_job_id,
                    hyperparameters,
                    metrics,
                    artifact_uri,
                    schema,
                    created_at,
                    promoted_at
                "#,
            )
            .bind(Uuid::now_v7())
            .bind(model_id)
            .bind(next_version_number)
            .bind(format!("autotune-v{next_version_number}"))
            .bind(job_id)
            .bind(best_hyperparameters)
            .bind(to_json(&execution.best_metrics))
            .bind(execution.best_artifact_uri.clone().or_else(|| {
                Some(format!(
                    "ml://models/{model_id}/versions/{next_version_number}"
                ))
            }))
            .bind(execution.best_schema.clone().unwrap_or_else(|| {
                let external_tracking =
                    interop::tracking_source_from_training_config(&training_config_for_schema);
                interop::normalize_model_version_schema(
                    Some(json!({
                        "signature": "tabular",
                        "engine": interop::effective_framework(&training_config_for_schema),
                        "objective_metric": objective_metric_name_for_schema.clone(),
                        "generated_by": "training-orchestrator",
                        "reproducibility": {
                            "dataset_ids": dataset_ids_for_schema.clone(),
                            "training_config": training_config_for_schema.clone(),
                            "hyperparameter_search": search_for_schema.clone()
                        }
                    })),
                    execution.best_artifact_uri.as_deref(),
                    Some(&training_config_for_schema),
                    None,
                    None,
                    external_tracking.as_ref(),
                )
            }))
            .fetch_one(&state.db)
            .await
            .map_err(|cause| db_error(&cause))?;

            best_model_version_id = Some(version_row.id);

            sqlx::query(
                "UPDATE ml_models SET latest_version_number = $2, current_stage = 'candidate', updated_at = NOW() WHERE id = $1",
            )
            .bind(model_id)
            .bind(next_version_number)
            .execute(&state.db)
            .await
            .map_err(|cause| db_error(&cause))?;
        }
    }

    let row = query_as::<_, TrainingJobRow>(
        r#"
        INSERT INTO ml_training_jobs (
            id,
            experiment_id,
            model_id,
            name,
            status,
            dataset_ids,
            training_config,
            hyperparameter_search,
            objective_metric_name,
            trials,
            best_model_version_id,
            submitted_at,
            started_at,
            completed_at,
            created_at
        )
        VALUES ($1, $2, $3, $4, 'completed', $5, $6, $7, $8, $9, $10, $11, $12, $13, $14)
        RETURNING
            id,
            experiment_id,
            model_id,
            name,
            status,
            dataset_ids,
            training_config,
            hyperparameter_search,
            objective_metric_name,
            trials,
            best_model_version_id,
            submitted_at,
            started_at,
            completed_at,
            created_at
        "#,
    )
    .bind(job_id)
    .bind(body.experiment_id)
    .bind(body.model_id)
    .bind(body.name.trim())
    .bind(to_json(&body.dataset_ids))
    .bind(resolved_training_config)
    .bind(search)
    .bind(objective_metric_name)
    .bind(to_json(&execution.trials))
    .bind(best_model_version_id)
    .bind(now)
    .bind(Some(now))
    .bind(Some(now))
    .bind(now)
    .fetch_one(&state.db)
    .await
    .map_err(|cause| db_error(&cause))?;

    Ok(Json(to_training_job(row)))
}

File: openfoundry-go/libs/ml-kernel-go/handlers/training.go
package handlers

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/openfoundry/openfoundry-go/libs/ml-kernel-go/domain/interop"
	"github.com/openfoundry/openfoundry-go/libs/ml-kernel-go/domain/training"
	"github.com/openfoundry/openfoundry-go/libs/ml-kernel-go/models"
)

// TrainingHandlers ports libs/ml-kernel/src/handlers/training.rs:
//   - GET  list_training_jobs
//   - POST create_training_job   (501 stub: chains
//                                 interop::merge_training_config_with_external
//                                 + training::execute_training; lands
//                                 with the libs/ml-kernel-go/domain/
//                                 interop port — 769 LOC of pure logic)
type TrainingHandlers struct {
	Pool *pgxpool.Pool
}

const trainingJobColumns = `id, experiment_id, model_id, name, status,
                            dataset_ids, training_config,
                            hyperparameter_search, objective_metric_name,
                            trials, best_model_version_id,
                            submitted_at, started_at, completed_at,
                            created_at`

func scanTrainingJob(s predictionsScanner) (models.TrainingJob, error) {
	var j models.TrainingJob
	var datasetIDsRaw, trainingConfigRaw, hyperSearchRaw, trialsRaw []byte
	var experimentID, modelID, bestVersionID *uuid.UUID
	var startedAt, completedAt *time.Time
	if err := s.Scan(
		&j.ID, &experimentID, &modelID, &j.Name, &j.Status,
		&datasetIDsRaw, &trainingConfigRaw, &hyperSearchRaw,
		&j.ObjectiveMetricName, &trialsRaw, &bestVersionID,
		&j.SubmittedAt, &startedAt, &completedAt, &j.CreatedAt,
	); err != nil {
		return j, err
	}
	j.ExperimentID = experimentID
	j.ModelID = modelID
	j.BestModelVersionID = bestVersionID
	j.StartedAt = startedAt
	j.CompletedAt = completedAt
	if len(datasetIDsRaw) > 0 {
		_ = json.Unmarshal(datasetIDsRaw, &j.DatasetIDs)
	}
	if j.DatasetIDs == nil {
		j.DatasetIDs = []uuid.UUID{}
	}
	if len(trainingConfigRaw) > 0 {
		j.TrainingConfig = trainingConfigRaw
	} else {
		j.TrainingConfig = json.RawMessage("{}")
	}
	if len(hyperSearchRaw) > 0 {
		j.HyperparameterSearch = hyperSearchRaw
	} else {
		j.HyperparameterSearch = json.RawMessage("{}")
	}
	if len(trialsRaw) > 0 {
		_ = json.Unmarshal(trialsRaw, &j.Trials)
	}
	if j.Trials == nil {
		j.Trials = []models.TrainingTrial{}
	}
	j.ExternalTraining = interop.TrackingSourceFromTrainingConfig(j.TrainingConfig)
	return j, nil
}

// ListTrainingJobs handles `GET /api/v1/training-jobs`.
func (h *TrainingHandlers) ListTrainingJobs(w http.ResponseWriter, r *http.Request) {
	rows, err := h.Pool.Query(r.Context(),
		`SELECT `+trainingJobColumns+` FROM ml_training_jobs
          ORDER BY submitted_at DESC, created_at DESC`)
	if err != nil {
		dbError(w, err)
		return
	}
	defer rows.Close()
	out := make([]models.TrainingJob, 0)
	for rows.Next() {
		j, err := scanTrainingJob(rows)
		if err != nil {
			dbError(w, err)
			return
		}
		out = append(out, j)
	}
	writeJSON(w, http.StatusOK, models.ListTrainingJobsResponse{Data: out})
}

// CreateTrainingJob handles `POST /api/v1/training-jobs`. Mirrors fn
// create_training_job verbatim: merges external tracking into the
// training config, runs ExecuteTraining (which picks the
// external-import / synthetic-trials / inline-records branch
// automatically), optionally registers a new model_version when
// auto_register_model_version + model_id + best_hyperparameters all
// resolve, then inserts the ml_training_jobs row with status='completed'.
func (h *TrainingHandlers) CreateTrainingJob(w http.ResponseWriter, r *http.Request) {
	var body models.CreateTrainingJobRequest
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	if strings.TrimSpace(body.Name) == "" {
		writeError(w, http.StatusBadRequest, "training job name is required")
		return
	}

	objectiveMetricName := derefString(body.ObjectiveMetricName, "accuracy")
	var search json.RawMessage
	if body.HyperparameterSearch != nil && len(*body.HyperparameterSearch) > 0 {
		search = *body.HyperparameterSearch
	} else {
		search = json.RawMessage("{}")
	}
	baseConfig := body.TrainingConfig
	if len(baseConfig) == 0 || string(baseConfig) == "null" {
		baseConfig = json.RawMessage(`{"engine":"tabular-logistic"}`)
	}
	resolvedConfig := interop.MergeTrainingConfigWithExternal(baseConfig, body.ExternalTraining)

	exec, err := training.ExecuteTraining(resolvedConfig, search, objectiveMetricName)
	if err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}

	now := time.Now().UTC()
	jobID := uuid.New()

	var bestModelVersionID *uuid.UUID
	if body.AutoRegisterModelVersion && body.ModelID != nil &&
		exec.BestHyperparameters != nil && len(exec.BestHyperparameters) > 0 {
		modelID := *body.ModelID
		var nextVersionNumber int32
		if err := h.Pool.QueryRow(r.Context(),
			`SELECT COALESCE(MAX(version_number), 0) + 1 FROM ml_model_versions WHERE model_id = $1`,
			modelID).Scan(&nextVersionNumber); err != nil {
			dbError(w, err)
			return
		}

		artifactURI := exec.BestArtifactURI
		if artifactURI == "" {
			artifactURI = "ml://models/" + modelID.String() + "/versions/" + itoaInt32(nextVersionNumber)
		}

		var schemaJSON json.RawMessage = exec.BestSchema
		if len(schemaJSON) == 0 {
			seed := map[string]any{
				"signature":        "tabular",
				"engine":           interop.EffectiveFramework(resolvedConfig),
				"objective_metric": objectiveMetricName,
				"generated_by":     "training-orchestrator",
				"reproducibility": map[string]any{
					"dataset_ids":            body.DatasetIDs,
					"training_config":        rawMessageOrNull(resolvedConfig),
					"hyperparameter_search":  rawMessageOrNull(search),
				},
			}
			seedRaw, _ := json.Marshal(seed)
			schemaJSON = interop.NormalizeModelVersionSchema(
				seedRaw, exec.BestArtifactURI, resolvedConfig, nil, nil,
				interop.TrackingSourceFromTrainingConfig(resolvedConfig),
			)
		}

		metricsJSON, _ := json.Marshal(exec.BestMetrics)

		var versionID uuid.UUID
		if err := h.Pool.QueryRow(r.Context(),
			`INSERT INTO ml_model_versions
                  (id, model_id, version_number, version_label, stage,
                   source_run_id, training_job_id, hyperparameters,
                   metrics, artifact_uri, schema, promoted_at)
                VALUES ($1, $2, $3, $4, 'candidate', NULL, $5, $6, $7, $8, $9, NULL)
                RETURNING id`,
			uuid.New(), modelID, nextVersionNumber,
			"autotune-v"+itoaInt32(nextVersionNumber),
			jobID, exec.BestHyperparameters, metricsJSON, artifactURI, schemaJSON,
		).Scan(&versionID); err != nil {
			dbError(w, err)
			return
		}
		bestModelVersionID = &versionID

		if _, err := h.Pool.Exec(r.Context(),
			`UPDATE ml_models SET latest_version_number = $2, current_stage = 'candidate', updated_at = NOW()
              WHERE id = $1`, modelID, nextVersionNumber); err != nil {
			dbError(w, err)
			return
		}
	}

	datasetIDsJSON, _ := json.Marshal(body.DatasetIDs)
	trialsJSON, _ := json.Marshal(exec.Trials)

	row := h.Pool.QueryRow(r.Context(),
		`INSERT INTO ml_training_jobs
              (id, experiment_id, model_id, name, status, dataset_ids,
               training_config, hyperparameter_search,
               objective_metric_name, trials, best_model_version_id,
               submitted_at, started_at, completed_at, created_at)
            VALUES ($1, $2, $3, $4, 'completed', $5, $6, $7, $8, $9, $10, $11, $12, $13, $14)
            RETURNING `+trainingJobColumns,
		jobID, body.ExperimentID, body.ModelID, strings.TrimSpace(body.Name),
		datasetIDsJSON, resolvedConfig, search, objectiveMetricName,
		trialsJSON, bestModelVersionID, now, now, now, now)
	j, err := scanTrainingJob(row)
	if err != nil {
		dbError(w, err)
		return
	}
	writeJSON(w, http.StatusOK, j)
}

func itoaInt32(n int32) string {
	if n == 0 {
		return "0"
	}
	var b [11]byte
	i := len(b)
	negative := false
	x := n
	if x < 0 {
		negative = true
		x = -x
	}
	for x > 0 {
		i--
		b[i] = byte('0' + x%10)
		x /= 10
	}
	if negative {
		i--
		b[i] = '-'
	}
	return string(b[i:])
}

func rawMessageOrNull(raw json.RawMessage) any {
	if len(raw) == 0 {
		return nil
	}
	var v any
	if err := json.Unmarshal(raw, &v); err != nil {
		return nil
	}
	return v
}

// loadTrainingJob is a small helper kept private but exported via
// the handler so future slices (eg. retry / cancel endpoints) can
// reuse it without re-implementing the column list.
func (h *TrainingHandlers) loadTrainingJob(ctx context.Context, id uuid.UUID) (*models.TrainingJob, error) {
	row := h.Pool.QueryRow(ctx,
		`SELECT `+trainingJobColumns+` FROM ml_training_jobs WHERE id = $1`, id)
	j, err := scanTrainingJob(row)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &j, nil
}
